### PR TITLE
[sysvabi64] Document requirements for tools wrt BTI

### PR DESCRIPTION
**Pull request by smithp35**
Add requirements for when a tool must generate a BTI instruction. This permits tools to elide BTI instructions when they can prove that no indirect branch to that location is possible from local information available to the tool.

Static linkers are not allowed to assume that all direct branch targets have a BTI instruction. If a veneer is required then the static linker must generate additional BTI instructions if needed.

A static linker is allowed to assume that a symbol that is exported to the dynamic symbol table has a BTI instruction.

In practice this will permit compilers to remove BTI instructions from static functions that do not have their address taken and that address escapes the function.

This matches the behavior of the GNU toolchain.

Fixes https://github.com/ARM-software/abi-aa/issues/196

### CODE REVIEW

**Comment by smithp35** [comment]
LLD patch to implement BTI aware thunks https://github.com/llvm/llvm-project/pull/108989

File: sysvabi64/sysvabi64.rst
Comment (smithp35) [reply sysvabi64/sysvabi64.rst:1702]: I'm thinking about how to write this without tying myself in knots. My current favourite is
"the address of that location is live and escapes out of the relocatable object." That should cover the case where the address of the is in a register or memory but there's no reader.

**Comment by smithp35** [comment]
Apologies for the force push, needed to rebase to pick up the new github actions.

File: sysvabi64/sysvabi64.rst
Comment (vhscampos) [inline sysvabi64/sysvabi64.rst:1693]: Typo "instuction"

File: sysvabi64/sysvabi64.rst
Comment (vhscampos) [inline sysvabi64/sysvabi64.rst:1696]: Consider putting each requirement in a list item.

File: sysvabi64/sysvabi64.rst
Comment (vhscampos) [inline sysvabi64/sysvabi64.rst:1717]: Typo: "bracnh"

File: sysvabi64/sysvabi64.rst
Comment (vhscampos) [inline sysvabi64/sysvabi64.rst:1710]: Can you please clarify why "if ... defined within the same link unit" is necessary? What happens in the case where the target is defined in another link unit?

File: sysvabi64/sysvabi64.rst
Comment (smithp35) [reply sysvabi64/sysvabi64.rst:1693]: ACK

File: sysvabi64/sysvabi64.rst
Comment (smithp35) [reply sysvabi64/sysvabi64.rst:1696]: ACK

File: sysvabi64/sysvabi64.rst
Comment (smithp35) [reply sysvabi64/sysvabi64.rst:1710]: I've gone with a more abstract escapes to an opaque entity. The original text refers to the boundary between the compiler/assembler and linker, but it can be more general than that. For example inline assembly that the compiler can't see into could indirect branch if the escaped address were on the input list.

File: sysvabi64/sysvabi64.rst
Comment (smithp35) [reply sysvabi64/sysvabi64.rst:1717]: ACK

**Review by smithp35 — COMMENTED** [review]
Thanks for the comments. I'll upload a new patch.

**Comment by vhscampos** [comment]
Ok by me. Thanks. But please wait for another approval.

**Review by statham-arm — APPROVED** [review]
I'm also happy with this now.

**Review by sallyarmneale — COMMENTED** [review]
Looks good to me

**Pull request merged**
